core/src/main/java/org/elasticsearch/ElasticsearchException.java

```diff
@@ -328,25 +319,94 @@ private void xContentHeader(XContentBuilder builder, String key, List<String> va
+            exception = new ElasticsearchException(throwable.getMessage(), throwable.getCause()) {
+                @Override
+                protected String getExceptionName() {
+                    return getExceptionName(throwable);
+                }
+            };
+        }
+        assert exception != null : "elasticsearch exception can't be null";
+        exception.toXContent(builder, params);
+    }
+
+    /**
+     * Render any exception as a xcontent, encapsulated within a field or object named "error". The level of details that are rendered
+     * depends on the value of the "detailed" parameter: when it equals to false, only a simple message based on the type and message of the
+     * exception is rendered. When it equals to true, all detail are provided including guesses root causes, cause and potentially stack
+     * trace.
+     */
+    public static void toXContentError(XContentBuilder builder, Params params, @Nullable Exception e, boolean detailed) throws IOException {
+        // No exception to render as an error
+        if (e == null) {
+            builder.field(ERROR, "unknown");
+            return;
+        }
+
+        // Render the exception with a simple message
+        if (detailed == false) {
+            String message = "No ElasticsearchException found";
+            Throwable t = e;
+            for (int counter = 0; counter < 10 && t != null; counter++) {
+                if (t instanceof ElasticsearchException) {
+                    message = t.getClass().getSimpleName() + "[" + t.getMessage() + "]";
+                    break;
+                }
+                t = t.getCause();
+            }
+            builder.field(ERROR, message);
+            return;
+        }
+
+        // Render the exception with all details
+        final ElasticsearchException[] rootCauses = ElasticsearchException.guessRootCauses(e);
+        builder.startObject(ERROR);
+        {
+            builder.startArray(ROOT_CAUSE);
+            for (ElasticsearchException rootCause : rootCauses) {
                 builder.startObject();
-                toXContent(builder, params, ex.getCause());
+                rootCause.toXContent(builder, new ToXContent.DelegatingMapParams(singletonMap(REST_EXCEPTION_SKIP_CAUSE, "true"), params));
                 builder.endObject();
             }
-            if (params.paramAsBoolean(REST_EXCEPTION_SKIP_STACK_TRACE, REST_EXCEPTION_SKIP_STACK_TRACE_DEFAULT) == false) {
-                builder.field(STACK_TRACE, ExceptionsHelper.stackTrace(ex));
-            }
+            builder.endArray();
         }
+        toXContent(builder, params, e);
+        builder.endObject();
+    }
+
+    /**
+     * Same as {@link ElasticsearchException#toXContentError(XContentBuilder, Params, Exception, boolean)} with all details.
+     */
+    public static void toXContentError(XContentBuilder builder, Params params, @Nullable Exception e) throws IOException {
```

Comment:
this name is not good. there is no *error* involved. I think you should maybe name it `generateFailureXContent` or somethign like this